src/RequestHandler/managedexports.cxx

```diff
@@ -43,6 +43,32 @@ http_get_raw_response(
     return pInProcessHandler->QueryHttpContext()->GetResponse()->GetRawHttpResponse();
 }
 
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+WCHAR*
+http_get_server_variable(
+    _In_ IN_PROCESS_HANDLER* pInProcessHandler,
+    _In_ PCSTR pszVariableName
+)
+{
+    PCWSTR pszVariableValue;
+    DWORD cbLength;
+    HRESULT hResult = pInProcessHandler
+        ->QueryHttpContext()
+        ->GetServerVariable(pszVariableName, &pszVariableValue, &cbLength);
+
+    if (hResult == S_OK && cbLength > 0 && pszVariableValue[0] != '\0')
+    {
+        WCHAR* pwszReturn = NULL;
```

Comment:
I personally prefer to return BSTR from these managed interop functions: https://msdn.microsoft.com/en-us/library/windows/desktop/ms221069(v=vs.85).aspx. They are type def'd to be WCHAR*. 

This function can also be simplified by calling SysAllocString().

```c++
EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
HRESULT
http_get_server_variable
(
    _In_ IN_PROCESS_HANDLER* pInProcessHandler,
    _In_ PCSTR pszVariableName,
    _Out_ BSTR* pwszRetVal
)
{
    PCWSTR  pszVariableValue;
    DWORD   cbLength;
    BSTR    pwzRetVal = NULL;
    HRESULT hr = pInProcessHandler
            ->QueryHttpContext()
            ->GetServerVariable(pszVariableName, &pszVariableValue, &cbLength);

    if (FAILED(hr) ||
        cbLength == 0 ||
        pszVariableValue[0] != L'\0') 
    {
        goto Finished;
    }

    *pwszRetVal = SysAllocString(pszVariableValue);

    if (*pwszRetVal == NULL)
    {
        hr = E_OUTOFMEMORY;
        goto Finished;
    }

Finished:

    return hr;
}
```

Reply:
Can you point me to the CLR code that does that? I wonder if SysAllocString uses CoTaskMemAlloc under the hood or if is compatible.

Reply:
I was reading this doc: https://docs.microsoft.com/en-us/dotnet/framework/interop/default-marshaling-behavior. I updated to use SysAllocString instead and it appears to be OK.